File: bestTimeToBuyAndSellStock.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
    	delta = [0] + [prices[i+1] - prices[i] for i in range(len(prices)-1)]
        p = delta
        for i in range(1, len(p)):
        	p[i] = max(p[i-1]+p[i], p[i])
        #print p
        return max(p)

s = Solution()
print s.maxProfit([5,4,3,1])

File: JAVA/bestTimeToBuyAndSellStock.java
public class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0, curProf = 0;
        int i = 1, difference;
        while (i < prices.length){
            difference = prices[i] - prices[i-1];
            while (curProf >= 0){
                curProf += difference;
                profit = (profit < curProf)? curProf: profit;
                if (i >= prices.length - 1 || curProf < 0)
                    break;
                i++;
                difference = prices[i] - prices[i-1];
            }
            curProf = 0;
            i++;
        }
        return profit;
    }
}